sysboundary/pml.cpp

```diff
@@ -0,0 +1,275 @@
+                pmlValue->at(fsgrids::pml::PGI2) = 1.0;
+                pmlValue->at(fsgrids::pml::PGI3) = 1.0;
+                pmlValue->at(fsgrids::pml::PFI1) = 1.0;
+                pmlValue->at(fsgrids::pml::PFI2) = 1.0;
+                pmlValue->at(fsgrids::pml::PFI3) = 1.0;
+                pmlValue->at(fsgrids::pml::PGJ2) = 1.0;
+                pmlValue->at(fsgrids::pml::PGJ3) = 1.0;
+                pmlValue->at(fsgrids::pml::PFJ1) = 1.0;
+                pmlValue->at(fsgrids::pml::PFJ2) = 1.0;
+                pmlValue->at(fsgrids::pml::PFJ3) = 1.0;
+                pmlValue->at(fsgrids::pml::PGK2) = 1.0;
+                pmlValue->at(fsgrids::pml::PGK3) = 1.0;
+                pmlValue->at(fsgrids::pml::PFK1) = 1.0;
+                pmlValue->at(fsgrids::pml::PFK2) = 1.0;
+                pmlValue->at(fsgrids::pml::PFK3) = 1.0;
+            }
+        }
+    }
+
+
+   Real xnum, xd;
+   Real xxn, xn;
+
+   // Attentuation Along the X-Dimension
+   #pragma omp parallel for collapse(3)
+   for (int kk = 0; kk < pmlDims[2] ; kk++){
+      for (int jj = 0; jj < pmlDims[1]; jj++){
+         for (int ii = 0; ii <pmlDims[0]; ii++){
+
+         // Get Global Arrays
+         pos=pmlGrid.getGlobalIndices(ii,jj,kk);
+
+         if (Xm==true){
+            if (pos[0]>=start && pos[0]<=P::pmlWidthX+start){
+               
+               // Get Local  Arrays
+               pmlValue = pmlGrid.get(ii, jj, kk);
+
+               xnum =P::pmlWidthX-pos[0] +start;
+               xd = P::pmlWidthX;
+               xxn =xnum/xd;
+               xn =0.33*(xxn*xxn*xxn);
+               pmlValue->at(fsgrids::pml::PGI2)=1/(1+xn);
+               pmlValue->at(fsgrids::pml::PGI3)=(1-xn)/(1+xn);         
+               xxn=(xnum-0.5)/xd;
+               xn=0.25*(xxn*xxn*xxn);
+               pmlValue->at(fsgrids::pml::PFI1)=xn;
+               pmlValue->at(fsgrids::pml::PFI2)=1/(1+xn);
+               pmlValue->at(fsgrids::pml::PFI3)=(1-xn)/(1+xn);
+            }
+         }
+
+
+         if (Xp==true){
+            if (pos[0]>=globalDims[0]-start- P::pmlWidthX && pos[0] <=globalDims[0]-start){
+            
+               // Get Local  Arrays
+               pmlValue = pmlGrid.get(ii, jj, kk);
+
+               xnum =start+ P::pmlWidthX-(globalDims[0]- pos[0]);
+               xd = P::pmlWidthX;               
+               xxn =xnum/xd;
+               xn =0.33*(xxn*xxn*xxn);
+               pmlValue->at(fsgrids::pml::PGI2)=1/(1+xn);
+               pmlValue->at(fsgrids::pml::PGI3)=(1-xn)/(1+xn);
+               xxn=(xnum-0.5)/xd;
+               xn=0.25*(xxn*xxn*xxn);
+               pmlValue->at(fsgrids::pml::PFI1)=xn;
+               pmlValue->at(fsgrids::pml::PFI2)=1/(1+xn);
+               pmlValue->at(fsgrids::pml::PFI3)=(1-xn)/(1+xn);
+               } 
+            }
+         }
+      }
+   }
+
+
+// Attentuation Along the Y-Dimension
+   #pragma omp parallel for collapse(3)
+   for (int kk = 0; kk < pmlDims[2] ; kk++){
```

Comment:
There is no reason to change the order of the loops so keep them k, j, i as this is more efficient for memory accesses. But then if it's anyway a threaded collapsed set of iterations it makes no difference until we run without threading, isn't it? Ping @ursg.

Reply:
Well I need to switch the order as in each loop I want to move along one dimension. I guess that when they get collapsed ,it will be fine.

Reply:
Two ways of answering here:
1. When you collapse it just makes a stack of single iterations (or groups thereof) without any ordering.
2. The code doesn't seem to have any dependence on the order (which would kill threadability anyway) so I don't understand our point or didnät see the dependence.

Reply:
I was wrong. There is no dependence. My bad. I will keep the same order .

Reply:
Good. :) Actually then it could be that it would be most efficient to have one single k-j-i threaded-collapsed loop, no?

Reply:
Sounds good. I will give it some more thought a bit later and do it however I don't think it is that important as this function is only executed once. I will give it some thought though, and implement it later today.